base/compiler/tfuncs.jl

```diff
@@ -980,27 +989,39 @@ end
     ⊑ = partialorder(𝕃)
 
     # If we have s00 being a const, we can potentially refine our type-based analysis above
-    if isa(s00, Const) || isconstType(s00)
-        if !isa(s00, Const)
-            sv = (s00::DataType).parameters[1]
-        else
+    if isa(s00, Const) || isconstType(s00) || isa(s00, PartialStruct)
+        if isa(s00, Const)
             sv = s00.val
+            sty = typeof(sv)
+            nflds = nfields(sv)
+            ismod = sv isa Module
+        elseif isa(s00, PartialStruct)
+            sty = s00.typ
+            nflds = fieldcount_noerror(sty)
```

Comment:
This appears to impose additional semantics on `PartialStruct` that are not clear to me are satisfied with the existing lattice elements. I'm thinking of cases like:

```
struct Foo
    x::Int
    y::Any
    Foo(x) = rand() ? new(1) : new(1, x)
end
```
It's not clear to me that the tmerge will respect the semantics you're imposing here. Perhaps it would be cleaner to give `PartialStruct` a new `ninitialized` field?

Reply:
> That said, this needs a comment somewhere to describe the semantics of a PartialStruct with short `fields`

Does c57d34c seem to address this?


> We should also audit all uses of `fields` and probably pkgeval this.

Yeah, I did audit all uses, but it would be a good idea to exercise with pkgeval.
@nanosoldier `runtests()`

> I'm not sure this logic is correct, because it doesn't look like it's actually looking at the length of `fields`, but I don't have the time to check right now.

Specifically regarding this logic, the length of `fields` is later referenced within `isdefined_tfunc`. `fieldcount_noerror` is used for iteration when applying `isdefined_tfunc` to all fields that might be uninitialized. However, that code path is hit only when bounds checking turned off and the field is not constant, making it quite niche.